src/expire.c

```diff
@@ -294,21 +298,33 @@ void activeExpireCycle(int type) {
             total_expired += data.expired;
             total_sampled += data.sampled;
 
-            /* Update the average TTL stats for this database. */
-            if (data.ttl_samples) {
-                long long avg_ttl = data.ttl_sum / data.ttl_samples;
-
-                /* Do a simple running average with a few samples.
-                 * We just use the current estimate with a weight of 2%
-                 * and the previous estimate with a weight of 98%. */
-                if (db->avg_ttl == 0) db->avg_ttl = avg_ttl;
-                db->avg_ttl = (db->avg_ttl/50)*49 + (avg_ttl/50);
-            }
+            /* If find keys with ttl not yet expired, we need to update the average TTL stats once. */
+            if (data.ttl_samples - origin_ttl_samples > 0) update_avg_ttl_times++;
 
+            repeat = data.sampled == 0 || (data.expired * 100 / data.sampled) > config_cycle_acceptable_stale;
             /* We can't block forever here even if there are many keys to
-             * expire. So after a given amount of milliseconds return to the
+             * expire. So after a given amount of microseconds return to the
              * caller waiting for the other active expire cycle. */
-            if ((iteration & 0xf) == 0) { /* check once every 16 iterations. */
+            if ((iteration & 0xf) == 0 || !repeat) { /* check once every 16 iterations or about to exit. */
+                /* Update the average TTL stats for this database, 
+                 * because this may reach the time limit. */
+                if (data.ttl_samples) {
+                    long long avg_ttl = data.ttl_sum / data.ttl_samples;
+
+                    /* Do a simple running average with a few samples.
+                    * We just use the current estimate with a weight of 2%
+                    * and the previous estimate with a weight of 98%. */
+                    if (db->avg_ttl == 0) {
+                        db->avg_ttl = avg_ttl;
+                    } else {
+                        for (int i = 0; i < update_avg_ttl_times; i++) {
```

Comment:
i wonder if there's a way to avoid this loop?

Reply:
Yes, it can convert to calculate the sum of a geometric progression. If we don't want a loop, we still need to use `pow` and double arithmetic.  Considering that `update_avg_ttl_times` is always less than 16 and the compiler may have some optimization, I don't know which is better. 
```
pow(0.98, update_avg_ttl_times) * (db->avg_ttl - avg_ttl) + avg_ttl
```

Reply:
good. we can compute the 16 constants ahead of time (static) and just use them here with a LUT instead of that loop

Reply:
OK. I also add some comments to explain the derivation process. Please take a look whether I need to add more.